leaf-server/src/main/java/org/dreeam/leaf/async/path/AsyncPath.java

```diff
@@ -78,211 +82,194 @@ public AsyncPath(@NotNull List<Node> emptyNodeList, @NotNull Set<BlockPos> posit
 
     @Override
     public boolean isProcessed() {
-        return this.processState == PathProcessState.COMPLETED;
+        return this.ready;
     }
 
     /**
-     * returns the future representing the processing state of this path
+     * Returns the future representing the processing state of this path
      */
-    public synchronized void postProcessing(@NotNull Runnable runnable) {
-        if (isProcessed()) {
+    public final void schedulePostProcessing(@NotNull Runnable runnable) {
+        if (this.ready) {
             runnable.run();
         } else {
-            this.postProcessing.add(runnable);
+            this.postProcessing.offer(runnable);
+            if (this.ready) {
+                this.runAllPostProcessing(true);
+            }
         }
     }
 
     /**
-     * an easy way to check if this processing path is the same as an attempted new path
+     * An easy way to check if this processing path is the same as an attempted new path
      *
      * @param positions - the positions to compare against
      * @return true if we are processing the same positions
      */
-    public boolean hasSameProcessingPositions(final Set<BlockPos> positions) {
+    public final boolean hasSameProcessingPositions(final Set<BlockPos> positions) {
         if (this.positions.size() != positions.size()) {
             return false;
         }
 
+        // For single position (common case), do direct comparison
+        if (positions.size() == 1) { // Both have the same size at this point
+            return this.positions.iterator().next().equals(positions.iterator().next());
+        }
+
         return this.positions.containsAll(positions);
     }
 
     /**
-     * starts processing this path
+     * Starts processing this path
+     * Since this is no longer a synchronized function, checkProcessed is no longer required
      */
-    public synchronized void process() {
-        if (this.processState == PathProcessState.COMPLETED ||
-            this.processState == PathProcessState.PROCESSING) {
-            return;
+    public final void process() {
+        if (this.ready) return;
+
+        synchronized (this) {
+            if (this.ready) return; // In the worst case, the main thread only waits until any async thread is done and returns immediately
```

Comment:
Question: does the main thread run `process()` when it _needs_ the path to be done? So, just purely by necessity, the main thread always has to wait at this point while the async thread is not done doing the processing?

Reply:
Because I assumed `awaitProcessing` would be called when the main thread actually needs the result.

If the main thread doesn't necessarily need the result when calling `process` yet, we can make an easy improvement here that avoids any waiting:

Replace `synchronized (this)` by `if (!hasStarted.get() && hasStarted.getAndSet(true) == false) {`
and add a new `AtomicBoolean hasStarted = new AtomicBoolean();` to this class.

The `synchronized` is then not necessary at all anymore as long as any thread that accesses any of the content checks `this.ready == true` at some time before that.